src/charm.py

```diff
@@ -1,19 +1,497 @@
+                    path=Config.get_license_path(license_name)
+                )
+                f = open(f"LICENSE_{license_name}", "x")
+                f.write(str(license_file.read()))
+                f.close()
+            except FileExistsError:
+                pass
+
+    @staticmethod
+    def _set_data_dir_permissions(container: Container) -> None:
+        """Ensure the data directory for mongodb is writable for the "mongodb" user.
+
+        Until the ability to set fsGroup and fsGroupChangePolicy via Pod securityContext
+        is available, we fix permissions incorrectly with chown.
+        """
+        for path in [Config.DATA_DIR, Config.LOG_DIR, Config.LogRotate.LOG_STATUS_DIR]:
+            paths = container.list_files(path, itself=True)
+            assert (
+                len(paths) == 1
+            ), "list_files doesn't return only the directory itself"
+            logger.debug(f"Data directory ownership: {paths[0].user}:{paths[0].group}")
+            if paths[0].user != Config.UNIX_USER or paths[0].group != Config.UNIX_GROUP:
+                container.exec(
+                    f"chown {Config.UNIX_USER}:{Config.UNIX_GROUP} -R {path}".split(" ")
+                )
+
+    def push_file_to_unit(
+        self,
+        parent_dir: str,
+        file_name: str,
+        file_contents: str,
+        container: Container = None,
+    ) -> None:
+        """Push the file on the container, with the right permissions."""
+        container = container or self.unit.get_container(Config.CONTAINER_NAME)
+        container.push(
+            f"{parent_dir}/{file_name}",
+            file_contents,
+            make_dirs=True,
+            permissions=0o400,
+            user=Config.UNIX_USER,
+            group=Config.UNIX_GROUP,
+        )
+
+    def unit_host(self, unit: Unit) -> str:
+        """Create a DNS name for a MongoDB unit.
+
+        Args:
+            unit_name: the juju unit name, e.g. "mongodb/1".
+
+        Returns:
+            A string representing the hostname of the MongoDB unit.
+        """
+        unit_id = unit.name.split("/")[1]
+        return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
+
+    # END: helper functions
+
+    # BEGIN: properties
+    @property
+    def mongos_initialised(self) -> bool:
+        """Check if mongos is initialised."""
+        return "mongos_initialised" in self.app_peer_data
+
+    @mongos_initialised.setter
+    def mongos_initialised(self, value: bool):
+        """Set the mongos_initialised flag."""
+        if value:
+            self.app_peer_data["mongos_initialised"] = str(value)
+        elif "mongos_initialised" in self.app_peer_data:
+            del self.app_peer_data["mongos_initialised"]
+
+    @property
+    def _unit_ip(self) -> str:
+        """Returns the ip address of the unit."""
+        return str(self.model.get_binding(Config.Relations.PEERS).network.bind_address)
+
+    @property
+    def is_external_client(self) -> Optional[str]:
```

Comment:
is this needed for mongos-k8s?

Reply:
ack - can I do in the follow up PR I am working on?